misc/docerina/src/main/java/org/ballerinalang/docgen/docs/BallerinaDocGenerator.java

```diff
@@ -226,23 +262,98 @@ private static void writeAPIDocs(ModuleLibrary moduleLib, Path output, boolean i
+                String zipFileURL = jsonResponse.get(JSON_KEY_FILE_URL).getAsString();
+                if (!Files.exists(sha256FilePath) || !Files.exists(zipFilePath)) {
+                    if (!docsDirPath.toFile().exists()) {
+                        Files.createDirectories(docsDirPath);
+                    }
+                    writeFileInCache(zipFileURL, sha256HashValue, zipFilePath, sha256FilePath);
+                } else {
+                    String hashValueInCache = Files.readString(sha256FilePath).trim();
+                    if (!sha256HashValue.equals(hashValueInCache)) {
+                        writeFileInCache(zipFileURL, sha256HashValue, zipFilePath, sha256FilePath);
+                    }
+                }
+                copyDocUIToProjectDir(output, zipFilePath);
+            } else {
+                throw new IOException("Response failed with status code: " + response.code());
             }
-        } else {
-            dest = output.resolve("index.html").toFile();
-            try {
-                FileUtils.copyInputStreamToFile(BallerinaDocGenerator.class
-                        .getResourceAsStream("/doc-ui/index.html"), dest);
-            } catch (IOException e) {
-                log.error("Failed to copy the doc ui.", e);
+        } catch (IOException e) {
+            if (Files.exists(zipFilePath)) {
+                String warning = """
+                        WARNING: Unable to fetch the latest UI from the central.
+                        This document is built using an existing version of the UI.
+                        """;
+                out.println(warning);
+                copyDocUIToProjectDir(output, zipFilePath);
+            } else {
+                log.error("Failed to copy the doc UI", e);
+            }
+        } finally {
+            client.dispatcher().executorService().shutdown();
+            client.connectionPool().evictAll();
+        }
+    }
+
+    private static void writeFileInCache(String fileURL, String hashValue, Path zipFilePath, Path hashFilePath)
+            throws IOException {
+        OkHttpClient client = new OkHttpClient();
+        Request request = new Request.Builder().get().url(fileURL).build();
+        try (Response response = client.newCall(request).execute()) {
+            if (response.code() == HTTP_OK && response.body() != null) {
+                ResponseBody responseBody = response.body();
+                byte[] contentInBytes = responseBody.bytes();
+                byte[] hash = BallerinaDocUtils.checkHash(contentInBytes, SHA256_ALGORITHM);
+                String checksum = BallerinaDocUtils.bytesToHex(hash);
+                if (checksum.equals(hashValue)) {
+                    Files.write(zipFilePath, contentInBytes);
+                    Files.write(hashFilePath, hashValue.getBytes());
+                } else {
+                    throw new IOException("Failed to download doc-ui zip file: File may be corrupted.");
```

Comment:
Change the log. "Failed to fetch API docs UI. UI Components may have been corrupted". We do not need to let the user know about how we get the API doc UI.

Reply:
Fixed in https://github.com/ballerina-platform/ballerina-lang/pull/42386/commits/fff5836b373e5e37d3680640581edf928a6344c0